rest-backend: sharing of configuration snippets

Namoshek opened the pull request:
This is a first prototype of my REST backend. It has been developed outside of the Elektra source tree and has just been integrated into CMake, so there might be lots of bugs and possible improvements (I'm quite sure the compiling of the tests can be massively improved because all source files are compiled again for each test, I just don't have a clue how to separate this). I'm aware of all compiler warnings; I developed with less flags, so these need to be resolved first. I'll also try to minimize the cryptlite library to what is used (or get rid of it by using boost directly).

When installing the tool, four files are copied into the elektra tool directory (elektra-rest-backend, run.sh, stop.sh and a config.js). Obviously the run.sh starts the application with the configuration of the config.js file. The service can also be tested with normal `make` (then everything remains in the directory).

To be able to use the convertion function, it is necessary to add some storage plugins to the following kdb-key `system/elektra/tools/kdbrest/formats`. This value should have a format in form of a string like `ini:ni xml:xmltool dump:dump`. Changes to this key require a restart of the REST service because the plugins are cached (on first use of the service).
It can also be a good idea to add individual mountpoints for `dir/repository` and `dir/users` (which can be changed in the config.hpp if necessary).

Important: The service caches all repository entries in-memory to increase performance (and that quite significantly). Direct changes to the database (without usage of the service) will only be visible after restart of the service.

The test suite is not complete yet.

---
# TODOs
- [x] Change `users` endpoint to `user`
- [x] Change format list of conversion endpoint
- [x] Change output of list for snippet conversions (add plugin name)
- [x] Fix issue: request raw format of snippet
- [x] Find better email regex
- [x] Remove unused warnings
- [x] Function documentation
- [x] Output of supported formats should be sorted by status of plugins
- [x] Find replacement for `cryptlite` lib

Namoshek commented:
I'm quite sure the build servers will need to install the [CppCMS](http://cppcms.com/wikipp/en/page/main) first (current version, 1.x.x).

markus2330 commented on cmake/Modules/FindCppCMS.cmake:
```diff
@@ -0,0 +1,14 @@
+if (CPPCMS_INCLUDE_DIR AND CPPCMS_LIBRARY)
+
+  # in cache already
+  SET(CppCMS_FOUND TRUE)
```
This looks weird, did you try to find a FindCppCMS.cmake in other projects? Maybe a robust solution exists already out there.

Namoshek replied:
Actually I copied this from another project out there and added something, because it didn't work without. But I agree, it is (and works) weird.

markus2330 commented on src/tools/rest-backend/CMakeLists.txt:
```diff
@@ -0,0 +1,38 @@
+find_package (CppCMS)
+find_package (Boost 1.50 COMPONENTS system filesystem thread REQUIRED)
```
Why is boost required? Is booster from cppcms not enough? Why do you need filesystem access?

Namoshek replied:
I have to create temporary files to use with the storage plugins for conversion. I can have another look at this and see if it works without boost too. I think I used it because boost offers methods to prevent creating already existing temp files.

Namoshek replied:
After thinking about it again, I can give you an answer. I'm using boost for synchronization (database cache) as well as the creation of temporary files. I tried using only the CppCMS package, but it didn't want to configure anymore (cmake). So I'll keep the import.

markus2330 commented on src/tools/rest-backend/CMakeLists.txt:
```diff
@@ -0,0 +1,38 @@
+find_package (Boost 1.50 COMPONENTS system filesystem thread REQUIRED)
+
+if (NOT CppCMS_FOUND)
+    message(SEND_ERROR "CppCMS package not found")
```
please use remove_tool at all places

Namoshek replied:
Can you elaborate on this one?

markus2330 replied:
Please do not use `SEND_ERROR`, but instead use `remove_tool` as done some lines below.

To make the removal of tools/bindings/plugins more visible is a general issue and should be solved once for all tools/bindings/plugins. Please create a new issue if you have some ideas about what a user would expect.

Btw. all warnings should be solved in the PR before further review, too.

Namoshek replied:
Ah ok, it didn't make sense to me that you wrote "there should be a big fat warning when `remove_*` is used" but at the same time insist on removing the warning. But now I got it and will remove it.

Warnings should be removed (see last commit some days ago). I used different settings when compiling locally (e.g. no -Wshadow).

markus2330 commented on src/tools/rest-backend/CMakeLists.txt:
```diff
@@ -0,0 +1,38 @@
+    
+    if (ENABLE_TESTING)
+	    add_subdirectory (tests)
+    endif(ENABLE_TESTING)
```
space missing

markus2330 commented on src/tools/rest-backend/Makefile:
```diff
@@ -0,0 +1,39 @@
+SOURCES		        = $(wildcard *.cpp)
```
Please remove the Makefile when CMake fully works

Namoshek replied:
Will keep it for testing as it increases development speed a lot. Final version will be cleaned up a bit.

markus2330 commented on src/tools/rest-backend/authentication_application.cpp:
```diff
@@ -0,0 +1,412 @@
+	}
+}
+
+void AuthenticationApp::setrank ()
```
setRank?

Namoshek replied:
I'm planning to add a very, very basic role-system with users, moderators and admins. So if this platform will ever be used, there is at least a way to edit and/or remove bad content. So this basically allows to set a rank.

markus2330 commented:
Yes, CppCMS will be installed soon. If something is not found you should remove the tool without error, though.

The exception handling seems to need some concept, or at least some explanation what the general idea is. There seems to be overuse of `try` blocks. Regions concerning security should be kept very short and marked. Did you look at the RAII concept?

markus2330 commented on src/tools/rest-backend/config.hpp:
```diff
@@ -0,0 +1,83 @@
+
+// Base path for the repository used by the rest service to store its data.
+// Path may NOT end with a slash.
+#define ELEKTRA_REST_CONFIG_REPOSITORY_PATH "dir/repository"
```
please prefer const variables (maybe in a class)

Namoshek replied:
The idea was that some of this values could be changed through CMake, especially paths.

markus2330 commented on src/tools/rest-backend/cryptlite/base64.h:
```diff
@@ -0,0 +1,198 @@
+/*
```
Why dont you use cppcms/base64.h??

Namoshek replied:
Will have a look!

Namoshek replied:
CppCMS is using a special base64 encoding adapted for use in URLs (some URL characters like `/` are not used). The CppCMS base64 class would probably be able to replace the cryptlite base64 functions, but I would still have to keep the cryptlite/base64.h for the cryptlite/sha256.h header. So I'm not sure if it is worth the effort.

As the cryptlite headers are a header-only interface for the boost crypt library anyway, I'm not sure if it is a viable option to try to replace it by Boost functions directly. It seems to me as if this would result in less readable code with exactly the same dependencies...

markus2330 replied:
The idea is to get rid of boost and cryptlite completely.

Why do you need sha256?

Namoshek replied:
As part of the JWTs (Json Web Tokens). They require the payload to be encrypted with an application secret, so that only the application can read and verify the cookie. I don't think there is another way around this. Using boost directly would bloat the code tremendously, which would result in refactoring into some cryptographic methods, what would lead to some sort of self-written cryptlite library. So I would like to keep it. I don't see a problem in using Boost anyway, because it is seen as part of the standard by most devs as far as I know.

markus2330 commented on src/tools/rest-backend/models.hpp:
```diff
@@ -0,0 +1,677 @@
+	kdb::KeySet _subKeys;
+};
+
+class PluginFormat
```
please use separate files

markus2330 commented on src/tools/rest-backend/models.hpp:
```diff
@@ -0,0 +1,677 @@
+	}
+
+private:
+	std::string _fileformat;
```
Please use `m_` as in other c++ code.

markus2330 commented on src/tools/rest-backend/root_application.cpp:
```diff
@@ -0,0 +1,356 @@
+{
+	if (request ().request_method () == "GET")
+	{
+            response().out() << "<h1>Welcome to the Elektra REST service!</h1>\n"
```
Any particular reason why templates are not used?

Namoshek replied:
No, I only added this part for the prototype so that it is clear which functions are available. I already planned so use a (maybe also better looking) template later. Was a quick & dirty solution in the evening.

Namoshek commented:
I always find it very weird if options are removed from the configuration without any notice. For example I wondered a quite significant amount of time why the qt-gui option is always removed when configuring. Is there another option to point out that the current configuration can't work without having to throw an error? It is ok to remove the tool from the list, but it seems completely wrong from usability perspective to do so without notifying the user...

Why do you think that try blocks are over-used or too long?
I use RAII explicitely in tests e.g.

markus2330 commented:
It looks like you have put a lot of effort. The main problem is the lack of documentation:

0. some general introduction + descriptions how to use it (including what you wrote here in the PR)
1. what did you want to achieve (use-cases)
2. architecture (how do you want to achieve it)
3. REST-interface

you can put everything in a single README.md, except the REST description should be in `doc` because it is shared with other implementations. If you want you can also write it in latex, we can discuss it tomorrow.

markus2330 commented:
There should be a big fat warning when `remove_*` is used.

Its mainly the density of how many `try` blocks are. If you have some policy to catch everything at some boarders it might be ok, but without architectural descriptions its hard to say.

Namoshek commented:
Will go through my code again and think about it. Maybe I can give an explanation, maybe it is just the way I code. Can't really tell right now. :)

Namoshek commented:
> you can put everything in a single README.md, except the REST description should be in doc because it is shared with other implementations. If you want you can also write it in latex, we can discuss it tomorrow.

I wrote a simple README.md draft (I'll add some more info when I'm sure that the service is final) and updated the CMake structure. The commit will follow as soon as I fixed some other stuff (don't want to trigger too many merge builds). I have already written a documentation for the REST interface in latex some time ago, but I don't really like the format yet (tables look too crowded), so I'll probably change it again. Or did you discuss otherwise on thursday (not to write it in latex) anyway?

Namoshek commented:
Implementing a template for the welcome page was hell of a hurdle. For some reason the linking didn't work the way it should (I guess because there are no real references in the code), but in the end it seems to work out.

markus2330 commented:
Build finished. No test results found.

Namoshek commented:
?
Isn't that because the fast merge doesn't run tests?

markus2330 commented:
The previous comment was written by the build server under my name. You are right, its because of the "fast" tests.

markus2330 commented:
About the REST documentation: the format does not really matter as long as we can read it.

Namoshek commented:
Ah ok, looked like you copied it trying to tell me something. Then nevermind... :)

markus2330 commented:
Currently the REST interface documentation is needed for coordination with @omnidan. As already said the goal is to have only one REST interface, implemented two times (maybe with some optional functionality for each implementation).

Maybe Markdown in the repo is the best format, though. It would be simple to read and write for everyone. We could also use some text format specifically designed for specifying REST interfaces.

Namoshek commented:
Please no own format creation for definition of REST interfaces... such work overhead isn't really viable. Markdown sounds ok. There is also software to create such documentation, but I'm not sure if this can be exported then.

But I still don't get why you think we need a common interface as our target group is completely different. @omnidan's service will be used by one or a bunch of admins, mine by (many) random people in public (at least that is the plan). Neither the public resource functions, nor the authentication can/should be shared... at least in my opinion.

Having a common way to respond to requests seems ok though, there are quite some points I'm not 100% sure about (like if it is ok/necessary to send the HTTP status explicitely as part of the response message additionally to the HTTP status that is available anyway).

markus2330 commented:
No, of course no new format, but one of the existing, e.g. http://swagger.io
Maybe @omnidan has some opinion about that, too.

I think the essence of the interfaces (get+set config with any config format) should be identical. I am sure there are more points to consider than error codes.

Namoshek commented:
Yes, I had swagger.io in mind, but I couldn't remember the name. It looks promising to me.

omnidan commented:
I haven't used swagger outside of the java world, so I'm not sure how nice it will work without it.

Although I have used [apiary](https://apiary.io/) before, it looks nice and uses a [markdown based syntax for writing api blueprints](https://apiblueprint.org/). There's also a [CLI tool](https://github.com/apiaryio/apiary-client) if we don't want to use their service.

I also found another api blueprint standard called [RAML](http://raml.org/).

markus2330 commented:
Looks like we should write a decision on this topic. Do you have any requirements on what the specification should be like? And which tools should be available for the specification?

From my side the only requirement is that the specification is human read/writeable (some integration in the rest of the docu would be nice, too) and the projects behind it are free software.

omnidan commented:
I agree. In addition to the requirements you already mentioned, having a markdown based syntax would be nice (would make it easier to integrate in the rest of the docs too).

markus2330 commented:
@Namoshek libcppcms-dev 1.0.5-1 is installed on the wheezy build agent. Currently no other debian repo exists, I asked the cppcms maintainer if he intends to provide one.

Namoshek commented:
I think the apiary CLI does look promising because it seems quite simple and easy to integrate with CMake.

@markus2330 Thank you! Personally I'm using a custom installation taken from the repository trunk on a ubuntu machine. Is there any reason you'd prefer an installation through the packet manager?

markus2330 commented:
Please do not forget to think about requirements for such a tool. Why is integration with CMake needed, do you want to generate code with it?

@Namoshek the reasons is reproducibility and the effort needed to do so on multiple machines with different OS each. If no packages can be provided we will have to do it anyway.

Namoshek commented:
No I don't want to use it for code generation, but I thought it would be useful for documentation creation (by using something like [Iglo](https://github.com/subosito/iglo)). Or did you have something else in mind (especially about documentation integration)?

Personally I don't have that much requirements. Of course it should be some easy readable documentation or be possible to generate some easily readable documentation. The separation of data modeling and API description (that API Blueprints offer) seems quite nice to me for example, but it would work without it too. But I don't think all that is mandatory. Human-readable formats are standard anyway and some doc generation seems also very common. So I don't know, but for me all of the suggested ones would work I guess.

@markus2330 got it!

markus2330 commented:
I created a new issue #917 for API documentation discussion.

Namoshek commented:
Currently I'm working on retrieving a list of installed storage plugins, so I can tell/show the user which formats are available (=offered by the REST service). Retrieving the plugin list and filtering storage plugins seems to work (I wrote a new function for the tools library to retrieve all plugins providing certain functionality; PR as soon as I'm sure it works as intended).

My problem is now that there doesn't seem to be a realiable way to find out the configuration format that is handled by a plugin. The format is sometimes the plugin name itself (e.g. `ini` plugin), sometimes it is in the `provides` list, but as there is no limitation of what can be inside the `provides` list, I cannot simply take "the next value != 'storage'", right? Is it possible to find a solution with current information or do I (still) need to specify some sort of enabled formats separately? @markus2330

markus2330 commented:
The problem is that there is no clear boundary where a configuration file format/syntax ends. E.g. there are hundred of incompatible INI formats out there. We had similar discussions in the qt-gui (it was about the file name extensions, e.g. see #175).

Please open a new issue for clarification of what exactly the contract means. (Maybe we extend the contract with an ontology for configuration file formats?)

Namoshek commented:
Ok, I think I should rebase.

markus2330 commented:
Yes, please rebase ;)

Namoshek commented:
A development preview can be found at http://namoshek.at:8080/ (and frontend preview on [port 9000](http://namoshek.at:9000)). Please be aware that the api description you see on the root is not up to date as it will be replaced by the apiary docu soon (I'm thinking about offering two docus - normally the rendered HTML and the blueprint if a parameter is set).

There are still open tasks that require other changes first, for a list see [tool/rest-backend project](https://github.com/ElektraInitiative/libelektra/projects/5).

Namoshek commented:
I think we'll need an auto-restart script for the backend as soon as it gets online, because not only `fstab` causes random crashes without further notice. I also experienced an issue with `simpleini`.

I added the converter function, feel free to check it out. But please do not use spaces in key-names. :smile:

markus2330 commented:
> I think we'll need an auto-restart script for the backend as soon as it gets online

The auto-restart would be also be useful now ;)
But please also collect crash dumps (or back traces should be enough).

> fstab /simpleini crashes:

fstab is old, unfinished and nearly untested, so no surprises there. For simpleini a recent change might be the problem.

Could you please create bug reports with back traces and the problematic snippets?

> I added the converter function, feel free to check it out. But please do not use spaces in key-names.

Very nice! Looks great! Sign up worked very well and also no big issues with converting. Also very nice to see the "report issue" functionality (the newlines seem to not work though) The popup on the right side in format conversions is visible a bit too short. It also says that you can contact us without any link (you can simply mention that the button can be pressed. And I would not use this red there, we want people to report issues!).

markus2330 commented:
I created an issue #1011. Please tell me if you need further feedback, I can also ask others.

Btw. the experience of the page is really nice, there is some fun into converting the snippets. Hopefully you can create the same experience for uploading snippets (seems to be down at the moment, the page only shows `{{ 'APP.NEWENTRY.PANEL.HEADING' | translate }}`).

Namoshek commented:
> The auto-restart would be also be useful now ;)

I see... I restarted it for now and will look for a solution later. I have to do some personal stuff in the morning first.

> (the newlines seem to not work though)

What browser have you been using? I tried with Chrome 53.0 on Windows and the current Firefox version (I think 45 or so) on Debian, it worked on both systems as expected. Is the pre-fill content URL encoded (in the address bar) when you use the function?

> It also says that you can contact us without any link (you can simply mention that the button can be pressed. 

The report function came later (and I didn't even know I was going to make it), so I'll change the text!

> seems to be down at the moment

Because you crashed the backend, the frontend cannot grab the available formats anymore from the backend (have to add a cache for them so they don't get fetched again and again). That's why the form isn't even loading. An error message when the backend doesn't respond is definitely needed though!

markus2330 commented:
> the location of kdb has changed to the system root

Do you mean the tool kdb or do you mean where files from the `dir` namespace are written to? For latter: yes, you need to write a script which only does `cd /mydir` before starting cppcms.

> (the newlines seem to not work though)

The issue only happend once, I could not reproduce it later. I thought you had fixed it, so I did not add it to #1011.

markus2330 commented on src/tools/rest-backend/README.md:
```diff
@@ -56,6 +56,31 @@ The REST service can also be configured. The file `rest-backend-config.js` in th
+
+To configure the rest-backend as service, it is possible to use `systemd` on most systems.
+
+1) Create a new service file under `/etc/systemd/system`, e.g. `touch /etc/systemd/system/rest-backend.service`.
```
Why first creating the file and then filling it with content later? What about using `cat` instead:
```
cat > /etc/systemd/system/rest-backend.service << HERE
[Unit]
...
HERE
```

Or do you want to explain that the folder `/etc/systemd/system` needs to be created first?

Namoshek replied:
The folder should be existing already if `systemd` is installed (quite sure this won't be the first service). But even if this is not the case, I guess the user will know what to do when getting an error because of a non-existing directory.

I changed it to your suggestion; didn't know `cat` supports heredoc.

markus2330 replied:
Heredoc is a shell feature, it works with every shell command.

markus2330 commented on src/tools/rest-backend/README.md:
```diff
@@ -56,6 +56,31 @@ The REST service can also be configured. The file `rest-backend-config.js` in th
+3) Reload the configuration of `systemctl` with `systemctl daemon-reload`.
+4) Enable the rest-backend service with `systemctl enable rest-backend.service`, a symlink should be created.
+5) Make sure the service is enabled with `systemctl is-enabled rest-backend.service`.
+6) Restart the rest-backen service with `systemctl restart rest-backend.service`. If everything went fine, the service should be reachable and `systemctl status rest-backend.service` should print information about the running service (PID, etc).
```
rest-backen?

Namoshek commented:
I replaced my own JWT implementation by the [libjwt](https://github.com/benmcollins/libjwt) implementation suggested and promoted on [jwt.io](https://jwt.io). I think this requires two new libraries to be installed on the server:
- [akheron/jansson](https://github.com/akheron/jansson)
- [benmcollins/libjwt](https://github.com/benmcollins/libjwt)

With replacing the cryptlite lib I'll wait until @petermax2 has moved his base64 and crypto stuff to a library, maybe I can reuse it.

markus2330 commented:
> maybe I can reuse it

Can you give some details what you need?

Namoshek commented:
I think the only three functions I would need a replacement for are `base64_encode`, `base64_decode` and `hs256_encode` (pseudo-names). For base64 I could probably also use CppCMS functions (that are URL functions, which shouldn't be a problem though), but if we have them in one of our libraries, why use another function. The crypto function is more interesting though...

markus2330 commented:
> but if we have them in one of our libraries, why use another function.

Seems like @petermax2 found another solution and it is not really necessary to expose base64 functions (which do not really fit in any of the existing libs). It is okay to use cppcms functionality.

> The crypto function is more interesting though...

You mean `hs256_encode`? @petermax2 Which of the libs support it and do you have an idea how we can expose them? Maybe via plugin contract?

> CppCMS https

The internal CppCMS webserver is for debugging only, so it will be used in an fcgi setup anyway.

Namoshek commented:
> Seems like @petermax2 found another solution and it is not really necessary to expose base64 functions (which do not really fit in any of the existing libs). It is okay to use cppcms functionality.

Ok, then I'll take the cppcms functions for that.

> You mean hs256_encode? @petermax2 Which of the libs support it and do you have an idea how we can expose them? Maybe via plugin contract?

Ok no, I was wrong. I need a function to encrypt passwords, so it should be sha256 instead of HS256, sorry. But I don't mind using another function that provides enough security (not md5 obviously).

> The internal CppCMS webserver is for debugging only, so it will be used in an fcgi setup anyway.

Ah ok.

Namoshek commented:
The last commit contains some significant changes regarding performance, thanks to some benchmarks. Can you have a look at the benchmarks @markus2330 and tell me if they look fine? I think they give enough info to write something about it; the only thing that needs to be done manually is defining multiple mount points.

Side note: Even with caching inserts take quite long. I think there is some space for improvements, but I'm not yet sure how I can do it without violating some (purely theoretical) assertions.

Namoshek commented:
After looking into it again and trying something, I don't think that there is a faster way to insert new data. Even with caching the whole KeySet retrieved from KDB, it is way slower than operating on a single entry.

markus2330 commented:
>  tell me if they look fine

Yes, certainly I will do. But I would prefer to directly cross check with your results/discussions.

> (purely theoretical) assertions.

Which ones?

> Even with caching the whole KeySet retrieved from KDB, it is way slower than operating on a single entry.

Can you elaborate here or point to which part in the benchmark you mean?

Namoshek commented:
> Yes, certainly I will do. But I would prefer to directly cross check with your results/discussions.

Ok, then I'll do some tests and write some stuff first.

> Which ones?

Well, as you know I'm using a cache for the storage service. I could theoretically only operate on the cache and omit checks on the database (@ create, update & delete; lookups are not of concern). Maybe this would increase speed (not even too sure tbh), but I don't know if this is a good idea.

> Can you elaborate here or point to which part in the benchmark you mean?

I'm talking about [this function](https://github.com/ElektraInitiative/libelektra/pull/829/files#diff-79b6c2f1448e83ecda59fd0a66430becR33) (and the same for users). As you can see, I'm doing first a check on the cache and afterwards I also have a look at the key database itself to ensure consistency. But because I need to store the new entry, I need the `kdb.get()` anyway.

What I tried is adding an additional cache for the KeySet retrieved in [the cache loading function](https://github.com/ElektraInitiative/libelektra/pull/829/files#diff-79b6c2f1448e83ecda59fd0a66430becR33), but because of its size I think it makes more sense to operate on single keys in above function.

Btw. I see a logic error in my code (adding the entry to the cache before checking the key database), I will fix that.

Namoshek commented:
I seem to have found another issue that I quite certainly introduced with one of the performance increasing changes.

I'm using `this->_kdb.get (ks, "dir/configs")` when building the cache and on-update I'm using `this->_kdb.get (ks, "dir/configs/path/to/some/snippet")`, which seems to not return the snippet (although it is definitely stored in the key database, I checked with kdb). Is there an explanation for this? If I use an own KDB instance per operation, it is very, very slow (especially inserts).

markus2330 commented:
Both issues seems to be a problem of your cachefilter, actually it should handle exactly these case. (return "dir/configs/path/to/some/snippet" even after "dir/configs" was retrieved, allowing kdbSet() of subparts).

Namoshek commented:
Well, the cachefilter plugin is not mounted currently. But I'm not using kdbSet on a subpart directly, I'm also using kdbGet on the subpart before, which seems to return an empty KeySet (i.e. `kdbGet(ks1, "dir/configs")` -> `kdbGet(ks2, "dir/configs/some/snippet")` -> `kdbSet(ks2, "dir/configs/some/snippet")`). Shouldn't kdbGet on a subpart at least return something?

markus2330 commented:
> Well, the cachefilter plugin is not mounted currently.

I think you should. If it works well, we will make its behavior as new default.

> Shouldn't kdbGet on a subpart at least return something?

Only if there was a change. On mountpoints where no update is needed, kdbGet() does nothing. It should be documented in the API docu, please open an issue if the docu is not clear on this matter.

Namoshek commented:
In the current state the cachefilter plugin doesn't solve the issue, so I'll probably update it.

markus2330 commented:
Thank you, that would be great! I think you now have enough experience to create a user experience someone would expect from Elektra: As you mentioned, kdbGet() should always refill the keyset with the parts requested and kdbSet() should also work with any subparts.

Namoshek commented on src/tools/rest-backend/service_storage.cpp:
```diff
@@ -0,0 +1,574 @@
+	// register read access
+	boost::shared_lock<boost::shared_mutex> lock (_mutex_entry_cache);
+
+	for (auto & elem : this->getAllEntriesRef ())
```
It is a vector, no KeySet. But with the cachefilter plugin this could be changed to a normal lookup then.

Namoshek commented on src/tools/rest-backend/benchmarks/benchmark_kdbrest_elektra.cpp:
```diff
@@ -0,0 +1,462 @@
+	timer.start ();
+
+	// the stuff to benchmark
+	std::vector<model::Entry> entries = service::StorageEngine::instance ().getAllEntries ();
```
The filter function does not return anything, so I have to store the vector somewhere to be able to reference it. Is there a way to avoid the copy constructor? (btw. are you sure it is even copied and not optimized away by RVO?)

By design I tried to keep the functions modular and to do not too many things in one function, especially because the REST interface offers a lot of options to modify the output.

markus2330 replied:
> Is there a way to avoid the copy constructor?

If you copy it from some internal vector from StorageEngine to the user: seems to be hardly avoidable.

> By design I tried to keep the functions modular

The design seems to be nice (I did not look into details though), the remarks are only about the benchmark.

Namoshek commented on src/tools/rest-backend/service_storage.cpp:
```diff
@@ -0,0 +1,574 @@
+model::Entry StorageEngine::getEntry (const std::string & key)
+{
+	// register read access
+	boost::shared_lock<boost::shared_mutex> lock (_mutex_entry_cache);
```
I thought iterating the entries could cause trouble if another request deletes one of them at the same time. But this would change with the cachefilter plugin anyway, as soon as it works.

Namoshek commented on src/tools/rest-backend/benchmarks/benchmark_kdbrest_mysql.cpp:
```diff
@@ -0,0 +1,621 @@
+	timer.stop ();
+
+	// print found entries
+	std::vector<model::Entry> entries = rsToEntryVector (rs);
```
You are right, this is a mistake. Definitely have to change and re-test this. Thank you for the good catch!

markus2330 commented:
Please remove "work in progress" if it is ready to review.

Namoshek commented:
I think changing the installation directory should be done similar to the frontend, but as I introduced a new CMake cache variable outside the tool directory, it would be better to first merge the frontend and rebase + change the backend afterwards. So installation directory is not final yet (minor change though).

markus2330 commented:
My idea was that I will only do a code review (without trying it out) and functionality is done with issues later (after it is merged).

Namoshek commented:
Is ok for me.

markus2330 commented on src/tools/rest-backend/model_user.hpp:
```diff
@@ -0,0 +1,298 @@
+                 */
+	kdb::KeySet::iterator addSubkeys (kdb::KeySet::iterator iter_start, kdb::KeySet::iterator iter_end)
+	{
+		kdb::Key & k = static_cast<kdb::Key &> (*this);
```
This looks weird, why not copy the Key?

markus2330 commented on src/tools/rest-backend/model_user.hpp:
```diff
@@ -0,0 +1,298 @@
+		}
+		catch (kdbrest::exception::SubkeyNotFoundException & e)
+		{
+			return 1; // default is user rank
```
should be documented above.

markus2330 commented on src/tools/rest-backend/model_user.hpp:
```diff
@@ -0,0 +1,298 @@
+	}
+	/**
+                 * Getter for the rank.
+                 * 3 = Admin
```
Please use `enum class` for Rank.

markus2330 commented on src/tools/rest-backend/model_user.hpp:
```diff
@@ -0,0 +1,298 @@
+		}
+	}
+
+	static bool less_than_username (User & l, User & r)
```
Why is this part of this class?

Namoshek replied:
Where would you place it? To me it makes sense that it is near the compared model.

markus2330 replied:
Usually you would have such functions standalone next to the class. But it is not really a big thing.

markus2330 commented on src/tools/rest-backend/rest-backend-config.json:
```diff
@@ -0,0 +1,14 @@
+	"http" : {
+		"script_names": ["/"]
+	},
+	"api_specification": {
```
Why only have this config here?

Namoshek replied:
What do you mean? It is an option that allows to set the target link for the api blueprint (raw and pretty).

markus2330 replied:
With elektrifying cppcms this would not matter anymore. It is okay to have it in the run-time config!

markus2330 commented on src/tools/rest-backend/root_application.cpp:
```diff
@@ -0,0 +1,308 @@
+	{
+		cppcms::json::value data;
+
+		data["elektra"]["version"] = KDB_VERSION;
```
This is the Elektra version build against, for the run-time values you need to query `system/elektra/version/constants`

markus2330 commented on src/tools/rest-backend/root_application.cpp:
```diff
@@ -0,0 +1,308 @@
+ * @brief helper method that attempts to parse post data as json
+ * @param request a request
+ * @return a json value containing the parsed post data
+ */
```
Please also document thrown exceptions.

markus2330 commented on src/tools/rest-backend/run.sh:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+./rest-backend -c rest-backend-config.json &
```
Please encode full paths here (use cmake constants and configure_file)

markus2330 commented on src/tools/rest-backend/run.sh:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+./rest-backend -c rest-backend-config.json &
+echo $! > rest-backend-running.pid
```
pids should be in /run

markus2330 commented on src/tools/rest-backend/service.hpp:
```diff
@@ -0,0 +1,139 @@
+         * It is designed as singleton, so all applications using it will share
+         * the same instance, which allows for further logic if necessary.
+         */
+class SearchEngine : public singleton<SearchEngine>
```
Do you really need singletons here?

markus2330 commented on src/tools/rest-backend/service.hpp:
```diff
@@ -0,0 +1,139 @@
+	void loadAllEntries ();
+	void loadAllUsers ();
+
+	std::vector<model::Entry> _entryCache;
```
Different style, please also use `m_` here.

markus2330 commented on src/tools/rest-backend/service_convert.cpp:
```diff
@@ -0,0 +1,317 @@
+	}
+	catch (kdbrest::exception::UnsupportedConfigurationFormatException & e)
+	{
+		throw e; // re-throw explicitely
```
Why not simply `throw;`

markus2330 commented on src/tools/rest-backend/service_convert.cpp:
```diff
@@ -0,0 +1,317 @@
+
+	// create temporary file used in communication with plugin
+	boost::filesystem::path tmpPath = boost::filesystem::unique_path ();
+	std::string tmpPathStr = "/tmp/" + tmpPath.string ();
```
Mixture of filesystem::path and strings.

markus2330 commented on src/tools/rest-backend/service_storage.cpp:
```diff
@@ -0,0 +1,606 @@
+		}
+	}
+
+	KDB kdb;
```
Do you really want to have a new `KDB` for every entry?

Namoshek replied:
No, but it doesn't work otherwise without the cachefilter yet. This is one of the things that can be changed then..

markus2330 replied:
Okay, then let us fix this separately.

markus2330 commented on src/tools/rest-backend/service_storage.cpp:
```diff
@@ -0,0 +1,606 @@
+
+} // namespace service
+
+} // namespace kdbrest
```
No newline here.

markus2330 commented on src/tools/rest-backend/singleton.hpp:
```diff
@@ -0,0 +1,22 @@
+#ifndef ELEKTRA_REST_TEMPLATE_SINGLETON_GUARD
+#define ELEKTRA_REST_TEMPLATE_SINGLETON_GUARD
+
+#include <assert.h>
```
Not used?

Please use ELEKTRA_ASSERT where needed.

markus2330 commented on src/tools/rest-backend/user_application.cpp:
```diff
@@ -0,0 +1,737 @@
+			// check permissions
+			if (user.getUsername () != username && user.getRank () < 100)
+			{
+				throw exception::InsufficientPermissionsException ();
```
Please avoid using exceptions as control flow.

markus2330 commented:
Please use separate commits to incorporate review work. Remove "work in progress" when everything has been dealt with. If you like the elektrification of cppcms/using run-time config can be done separately (other PR): then please open an issue for it.

Namoshek commented on src/tools/rest-backend/.gitignore:
```diff
@@ -0,0 +1,3 @@
+CMakeCache.txt
+nbproject/
```
nbproject is for project files from netbeans. If we do not have it elsewhere in our gitignores, we should keep it.

No, I did not use in-source builds for the backend. Why?

markus2330 replied:
> nbproject

I meant to simply add a comment next to the line.

> No, I did not use in-source builds for the backend. Why?

This was suggested by ignoring CMakeCache.txt.

Namoshek replied:
CMakeCache.txt appeared in the source for some reason, maybe netbeans tries to compile something, I don't really know. I added comments to the gitignore.

Namoshek commented on src/tools/rest-backend/README.md:
```diff
@@ -0,0 +1,99 @@
+Following options are available at compile time:
+- `REST_REPOSITORY_CONFIGS` representing a path to the location where configuration snippets should be stored. An example would be `-DREST_REPOSITORY_CONFIGS=dir/configs`. The default value is `dir/configs`.
+- `REST_REPOSITORY_USERS` representing a path to the location where user information should be stored. An example would be `-DREST_REPOSITORY_USERS=user/rest/service/users`. The default value is `dir/users`.
+- `REST_AUTH_ENCRYPT_KEY` holding an encryption key to be used to encrypt authentication tokens. This option is mandatory because the default value is not secure. It can be a simple non-empty string like `a_5ecure-encrypt1on_key`.
```
Then it must be one-time generated, e.g. during cmake build.

Namoshek commented on src/tools/rest-backend/README.md:
```diff
@@ -0,0 +1,99 @@
+## Run and Configure ##
+
+To run the `rest-backend` we need to find out where it has been installed to. This can be done by running the command `kdb list-tools` which will output a list of installed tools and something like `External tools are located in /usr/local/lib/elektra/tool_exec`. With this path we can run the service like:
+`cd /usr/local/lib/elektra/tool_exec/ && sh run-rest-backend.sh`
```
Does it work as shell script if I omit the .sh?

Namoshek commented on src/tools/rest-backend/authentication_application.cpp:
```diff
@@ -0,0 +1,307 @@
+/**
+ * @brief handler for the authentication resource.
+ *		  attempts an authentication with the submitted credentials. in case the
+ *		  attempt succeeds, a session token will be returned, otherwise an error.
```
That's not my indentation, but from the clang reformatter. I think the docu errors are because I changed something, will fix it.

Namoshek commented on src/tools/rest-backend/authentication_application.cpp:
```diff
@@ -0,0 +1,307 @@
+		}
+		catch (kdbrest::exception::InvalidPostDataFormatException & e)
+		{
+			RootApp::setBadRequest (response (), "The submitted data is not of type application/json.",
```
What do you mean?

markus2330 replied:
At some places exceptions are abused for control flow. If you could avoid at least this it would be better. Your style is very exception-centric, which would be hard to change now. Hopefully they are not too expensive, did you benchmark it?

Namoshek replied:
Ok now I got it, thanks. You noticed it somewhere else already, so I thought you might mean something else. I will try to change where it is clearly abused for control flow.

No, I did not benchmark exception stuff. Exceptions are also used more in the controllers than in the services (there they are only thrown). In the benchmarks I assume to not receive any exceptions at all (no exception handling implemented there).

Namoshek commented on src/tools/rest-backend/authentication_application.cpp:
```diff
@@ -0,0 +1,307 @@
+		// find username
+		try
+		{
+			username = requestData.get<std::string> ("username");
```
If I do it in a subfunction, how can I know when to return early (because an error was set)?

markus2330 replied:
You could let exceptions through the subfunction, then you catch exceptions and use `setBadRequest`.

And it is not necessarily about saving lines, but about having functions that one can easily read and understand.

markus2330 replied:
You could let exceptions through the subfunction, then you catch exceptions and use `setBadRequest`.

And it is not necessarily about saving lines, but about having functions that one can easily read and understand.

Namoshek replied:
I changed the retrieval of inputs to something different in places where it makes sense (3 or more input fields), but for two fields it isn't viable, so I'd like to keep it that way here.

Namoshek commented on src/tools/rest-backend/authentication_application.cpp:
```diff
@@ -0,0 +1,307 @@
+		{
+			RootApp::setInternalServerError (response (), "Something went wrong while creating the session token.",
+							 "AUTH_CREATE_TOKEN_ERROR");
+			::jwt_free (jwt); // free memory
```
Good catch, thank you!

Namoshek commented on src/tools/rest-backend/authentication_application.cpp:
```diff
@@ -0,0 +1,307 @@
+						std::string orUser)
+{
+#ifdef ELEKTRA_REST_NO_AUTH
+	return true;
```
Not needed, will delete it. Early in development I thought it might be useful, but it isn't.

Namoshek commented on src/tools/rest-backend/authentication_application.hpp:
```diff
@@ -0,0 +1,52 @@
+#ifndef ELEKTRA_REST_AUTHENTICATIONAPP_HEADER_GUARD
```
Standard headers?

markus2330 replied:
Header in the sense of the head of the file. See at the tail of doc/CODING.md

Namoshek commented on src/tools/rest-backend/conversion_application.cpp:
```diff
@@ -0,0 +1,198 @@
+
+	if (request ().request_method () == "POST")
+	{
+		// check if request data is of type application/json
```
Why?

markus2330 replied:
The comments are "executable" then and one can view them while the code is executed. You know, the poor man's debugger ;)

Namoshek replied:
Well ok, but here we get an immediate result via the http response... so I can see if it is executed or not anyway. If I start introducing a logger for this small thingy, I can do it for the whole project - and that would be quite some effort...

markus2330 replied:
It is okay if you use assert+logs only where they are most urgent (and useful). They can be added at other places later (if needed, usually people tend to overuse logs anyway).

markus2330 replied:
It is okay if you use assert+logs only where they are most urgent (and useful). They can be added at other places later (if needed, usually people tend to overuse logs anyway).

Namoshek commented on src/tools/rest-backend/create_entries.sh:
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+for i in $(seq 100 150)
```
Unimportant script to create some dummy data, I think it can be removed or replaced by some more useful test data as you suggested.

Namoshek commented on src/tools/rest-backend/crypto.cpp:
```diff
@@ -0,0 +1,32 @@
+namespace crypto
+{
+
+bool sha256_encrypt (unsigned char * input, unsigned long length, unsigned char * output)
```
Interesting suggestion, not sure why I haven't thought of that.

Namoshek commented on src/tools/rest-backend/database_application.cpp:
```diff
@@ -0,0 +1,902 @@
+		return; // quit early, error response already set
+	}
+
+	model::Entry newEntry ("");
```
This one is actually only 50 lines long, while our coding standard is fine up to 100 lines?!

markus2330 replied:
Seems like you like long functions ;) I did not count the lines, but it seemed for me that creating the model in another function would be natural.

Namoshek commented on src/tools/rest-backend/database_application.cpp:
```diff
@@ -0,0 +1,902 @@
+	// retrieve submitted data
+	if (keyName.empty ())
+	{
+		try
```
This one really is, yes. :smile:

Namoshek commented on src/tools/rest-backend/model_user.hpp:
```diff
@@ -0,0 +1,298 @@
+                 * It is an alternative for an upwards cast.
+                 * @param key A kdb::Key object
+                 */
+	inline User (kdb::Key & k) : kdb::Key (k)
```
Interesting to know, wasn't aware of that! Is it ok expressing it explicitely or should I remove it?

Namoshek commented on src/tools/rest-backend/CMakeLists.txt:
```diff
@@ -0,0 +1,58 @@
+    set (REST_AUTH_ENCRYPT_KEY "please_change_to_secret_key" CACHE STRING "Encryption key to be used for auth tokens")
+    set (REST_JWT_ISSUER "additional_secret_issuer" CACHE STRING "The issuer to be used for JWT tokens, should be kept secret")
+    set (REST_REPOSITORY_CONFIGS "dir/configs" CACHE STRING "Path to be used to store configuration snippets")
+    set (REST_REPOSITORY_USERS "dir/users" CACHE STRING "Path to be used to store user information")
```
You are right, it would make more sense to have them run-time configured. Would it make sense to also add them to the configuration file provided by cppcms (they are tightly-coupled anyway)? Is it sufficient to mount the cppcms configuration or do you have something else in mind?

markus2330 replied:
The best way would be full integration. For that we would only need a KeySet to cppcms::json::value transformation. Then you could start the service with the constructor `service (json::value const &v)`.

As byproduct we easily could have a plugin that can write json files!

Namoshek replied:
That doesn't sound very trivial and a cppcms specific json plugin doesn't seem a good idea either (especially because we have a json plugin already). (If I understood right...)

I'll think this through and see what I can do.

Namoshek commented on src/tools/rest-backend/authentication_application.cpp:
```diff
@@ -0,0 +1,307 @@
+		}
+
+		/* authentication was successful, so lets build the token */
+		::jwt_t * jwt;
```
How do I invoke `jwt_new` with a smart pointer (function takes pointer to a pointer)?

As the jwt is initialized after the pointer is created, how do I ensure the deleter does not get called if it was not initialized properly?

markus2330 replied:
> How do I invoke jwt_new with a smart pointer (function takes pointer to a pointer)?
> As the jwt is initialized after the pointer is created, how do I ensure the deleter does not get called if it was not initialized properly?

Simply call `jwt_new` and after checking that it worked initialize the `unqiue_ptr` (raw pointer as parameter). So something like:

```c++
jwt_t * jwt;
jwt_new (jwt);
if (!jwt) { return; } // do something...
std::unique_ptr<jwt_t, void(*)(jwt_t*)> mjwt (jwt, jwt_free);
```

Namoshek commented:
First part of the changes, should make it a lot easier to see what is still open.

markus2330 commented:
Does github provide some way to approve individual changes? If not I would use emoji to mark them. But I think I will have to go through everything again anyway.

Namoshek commented:
I asked myself the same, as it would be a lot easier for me to keep an overview. But emojis are fine as well if you don't find an option...

markus2330 commented:
A dedicated feature would be better, then you could have a progress bar and an automatical approval after every individual message is approved. But it will also work with emoji.

Do you prefer thumbs up for things that are resolved or thumbs down for what is missing ;)

I think the way it is supposed to work is to start a second review and comment again the messages which are still not resolved. Then approve the first review. That already worked for #891 (which was also rather large, but I think not as large as this one).

Namoshek commented:
> A dedicated feature would be better, then you could have a progress bar and an automatical approval after every individual message is approved.

That doesn't work with every requested change. For example the new lines at the end of files - I added them already, but github thinks you criticised something else. Or when I add a requested addition in another place (e.g. one paragraph above), it doesn't work either.

I don't really mind how you do it, as long as I can see that it is ok. ^^

markus2330 commented:
> That doesn't work with every requested change.

No, I did not think about something automatic, but where someone (maybe even you) click away every comment that was dealt with. Nevertheless, the critique is on high level, the code review is done quite nicely.

> I don't really mind how you do it, as long as I can see that it is ok. ^^

I am sure we will do fine ;) The main problem is the coordination when to start the review and when to merge: and I think we solved that nicely with labels.

Namoshek commented:
> No, I did not think about something automatic, but where someone (maybe even you) click away every comment that was dealt with.

Yeah, I thought about something semi-automatic. Currently github seems to recognize if something was changed, but only if the mentioned line was changed. But as you have to go through everything manually anyway...

One of the last open tasks for me is to simplify the very long functions, then everything should be covered (I got quite some more commits that are not pushed yet). I'll tell you when you can do a second review.

Namoshek commented:
I hope my Elektra integration makes sense!

Namoshek commented:
Somehow I can't get `systemctl` to work anymore - server won't start. Starting it normally works perfectly. I know that it will not work with the example given in the readme because of `kdb` being called without absolute path, but even with path it doesn't seem to work. Think this requires further investigation... help welcome!

markus2330 commented on src/tools/rest-backend/service_config.cpp:
```diff
@@ -0,0 +1,186 @@
+namespace service
+{
+
+const std::string REGEX_CONF_KEY = "((?:[a-zA-Z_]+(?:\\.?[a-zA-Z_])+)?)\\.?#_?([0-9]+)\\.?((?:[a-zA-Z0-9_#]+(?:\\.?[a-zA-Z0-9_#])+)?)";
```
The regex is quite heavy and nevertheless it seems to not do escaping correctly. Why not iterate the name and avoid this regex? Btw. the name is confusing too, isn't it to match arrays?

Namoshek replied:
Yes it is purely for arrays. I tried all combinations that came to my mind with an online regex tester and it seems to work properly. The whole transformation was probably the most complicated stuff...

markus2330 replied:
> Yes it is purely for arrays. I tried all combinations that came to my mind with an online regex tester and it seems to work properly. The whole transformation was probably the most complicated stuff...

Okay, then please give it a better name and add some test cases.

Namoshek replied:
Maybe I'll have a look at iterating keys, but somehow the whole transformation seems complicated.

markus2330 commented on src/tools/rest-backend/service_config.cpp:
```diff
@@ -0,0 +1,186 @@
+		{
+			try
+			{
+				config[array_index];
```
Why trying to wrongly access it? Why not asking for its length?

markus2330 commented on src/tools/rest-backend/service_config.cpp:
```diff
@@ -0,0 +1,186 @@
+		// otherwise we can set directly
+		try
+		{
+			if (std::to_string (key.get<int> ()).length () == key.get<std::string> ().length ())
```
What is this comparision for?

Namoshek replied:
CppCMS has the setting "service.ip" ... as I don't really know the value type, I tried to parse it as integer first and this destroys the IP. It is a dirty fix to ensure only values that are really integers also get read as such.

markus2330 replied:
So actually `key.get<int>` is buggy? Does 239b08722b87d57346b73eaba9b40ddda547b94c fix this problem?

Namoshek replied:
Yes, this seems to fix the problem. But now I'm not sure if I should rebase, because it might kill the review. For the moment I only changed that one line in my code. (but I can merge it when rebasing after the review too)

markus2330 commented on src/tools/rest-backend/service_config.cpp:
```diff
@@ -0,0 +1,186 @@
+		}
+		catch (kdb::KeyTypeConversion & e)
+		{
+			// do nothing, it's fine
```
Why?

Namoshek replied:
Because it doesn't matter... I'm just checking whether it is an integer, if not I try the next data type (which is a string in this case).

markus2330 commented on src/tools/rest-backend/service_config.cpp:
```diff
@@ -0,0 +1,186 @@
+		}
+		try
+		{
+			std::string val = key.getString ();
```
Why not use `.get<bool>()`?

Namoshek replied:
Will that not make almost everything `true`?

markus2330 replied:
Why should it? 2f18998538f65f8a82b26e430f98dfa9834b4d70

Btw. to make it accept more values, we have the boolean plugin.

Namoshek replied:
Ok, but now it gets interesting: How do I know whether it is a bool or an integer?

markus2330 replied:
We cant know without metadata (you also cannot know if "true" is a string or boolean). The specification with spec plugin copies metadata to keys (e.g. metadata `type`).

Namoshek replied:
We don't have a specification for cppcms and I don't really want to write one to be honest.

CppCMS doesn't seem capable of converting 1 and 0 into true and false or the other way round, it gives bad_cast on every conversion attempt. So I'll have to assume that there won't be integers with values 1 and 0, I guess.

markus2330 commented on src/tools/rest-backend/service_config.cpp:
```diff
@@ -0,0 +1,186 @@
+			}
+			return;
+		}
+		catch (kdb::KeyTypeConversion & e)
```
How can this happen if you use `getString()`?

markus2330 commented on src/tools/rest-backend/service_config.cpp:
```diff
@@ -0,0 +1,186 @@
+		try
+		{
+			std::string val = key.getString ();
+			if (val == "true")
```
code duplicate

markus2330 commented on src/tools/rest-backend/user_application.cpp:
```diff
@@ -0,0 +1,734 @@
+	// delete user
+	else if (request ().request_method () == "DELETE")
+	{
+		try
```
why not split to separate method?

markus2330 commented on src/tools/rest-backend/.gitignore:
```diff
@@ -0,0 +1,6 @@
+CMakeFiles/
+
+#ignore netbeans project files
+nbproject/
```
newline missing at end of file

Namoshek replied:
Made non intentionally, as other .gitignores don't have one either...?!

Namoshek replied:
Weird, in my code version it definitely has none (maybe it was added recently, there were some changes if I remember correctly). I can add nbproject/ also there.

markus2330 commented on src/tools/rest-backend/singleton.hpp:
```diff
@@ -0,0 +1,28 @@
+	}
+
+protected:
+	singleton ()
```
Protect copy-constructor and move-constructor, too?

markus2330 commented on src/tools/rest-backend/service_convert.cpp:
```diff
@@ -0,0 +1,326 @@
+		{
+			if (boost::starts_with (provider, "storage/"))
+			{
+				format = provider.substr (8);
```
avoid magic constants. Use something like `const char storage [] = "storage/"; const size_t sizeOfStorage = sizeof(storage)-1;` (if you meant that).

markus2330 commented on src/tools/rest-backend/service.hpp:
```diff
@@ -0,0 +1,176 @@
+		this->m_enabledFormats = loadEnabledFormats ();
+	}
+
+	std::vector<kdbrest::model::PluginFormat> & getEnabledFormats ()
```
try to also provide const versions

markus2330 commented on src/tools/rest-backend/service.hpp:
```diff
@@ -0,0 +1,176 @@
+	std::vector<kdbrest::model::PluginFormat> loadEnabledFormats ();
+	model::PluginFormat findSuitablePlugin (const std::string & format);
+
+	kdbrest::model::ConfigFormat exportTo (const std::string format, model::Entry & entry);
```
Isn't this const?

markus2330 commented on src/tools/rest-backend/service.hpp:
```diff
@@ -0,0 +1,176 @@
+	bool updateEntry (model::Entry & entry);
+	bool deleteEntry (model::Entry & entry);
+
+	bool entryExists (const std::string & key);
```
const?

Namoshek replied:
Will probably work now, but as soon as I start using a global KDB instance again (for cachefilter later), it will not work.

markus2330 commented on src/tools/rest-backend/root_application.hpp:
```diff
@@ -0,0 +1,99 @@
+public:
+	RootApp (cppcms::service & srv);
+
+	virtual void welcome ();
```
const?

markus2330 replied:
Ok!

markus2330 commented on src/tools/rest-backend/model_user.hpp:
```diff
@@ -0,0 +1,355 @@
+	 * 
+     * @param key A kdb::Key object
+     */
+	inline User (kdb::Key & k) : kdb::Key (k)
```
useless inline (constructor/methods are implicitly inline when directly implemented in the class)

markus2330 commented on src/tools/rest-backend/config.cpp:
```diff
@@ -0,0 +1,152 @@
+/* set dummy variables for statics */
+
+// JWT settings
+std::string Config::jwt_encryption_key = std::string ();
```
please avoid globals, if necessary make Config a singleton, (1 global is better than n globals)

Namoshek replied:
Why n globals? Config shouldn't be initialized on static access? Or am I wrong and I should use a private constructor to prevent that?

markus2330 replied:
Currently every setting such as jwt_encryption_key is a class variable and could be a object variable of a singleton.

Namoshek replied:
Well that's obviously correct, but what is the advantage of making them member variables of a singleton? I only see an disadvantage and that is a more complicated call: `Config::instance().jwt_encryption_key` instead of `Config::jwt_encryption_key`...

markus2330 replied:
There would be a small advantage with getter/setters.

Namoshek replied:
You are basically right, I could also use the json::value. But at some point I need to validate the configuration anyway (on startup).

But it doesn't seem perfect either way; cppcms tools have weird behavior here and there. The worst is that you can use `json::value::set('this.are.objects.with.a.variable', true)` to set nested objects, but you can't use the same for arrays. I tried:
```
json::value::set('gimme.array[0]', true)
json::value::set('gimme.array.[0]', true)
json::value::set('gimme.array.0', true)
json::value::set('gimme.array.#0', true)
```
If any of these did work, the whole transformation would be so much easier...

Namoshek replied:
Btw. instead of changing configuration access to getters, I can also access the json::value... doesn't really change that much.

markus2330 commented on src/tools/rest-backend/config.cpp:
```diff
@@ -0,0 +1,152 @@
+ * @param config the configuration to use
+ * @return 0 in case of success, 1 otherwise
+ */
+int Config::initializeConfiguration (cppcms::json::value & cfg)
```
Why not return bool? And do you really modify `cfg`?

markus2330 commented on src/tools/rest-backend/config.cpp:
```diff
@@ -0,0 +1,152 @@
+
+	// clang-format on
+
+	if (error)
```
Why not `return error;`?

markus2330 commented on src/tools/rest-backend/config.cpp:
```diff
@@ -0,0 +1,152 @@
+	
+	// JWT settings
+	try {
+		Config::jwt_encryption_key = cfg.get<std::string> ("backend.jwt.encryption_key");
```
Why not fetch from KeySet here?

Namoshek replied:
As everything is already transformed, it seemed easier and more consistent this way. Not a big deal to change it I guess, but I'm glad about everything that works for the moment. :)

markus2330 replied:
Exactly, and "-" should be avoided, too. And as last part there should be a profile. Maybe `/sw/elektra/snippets/#0/current`?

Namoshek replied:
Is the profile something dynamic or how do I have to understand it? Because dynamic profiles and stuff seems a bit overkill for such a service.

About the "-"... I attempted to use the tool name, but in this case I need to use something static I guess?

markus2330 replied:
> Is the profile something dynamic or how do I have to understand it? Because dynamic profiles and stuff seems a bit overkill for such a service.

It is very simple, but useful: If we later implement command-line parameter we would add `--profile` or `-P` which allows to change the value from `current` to something else. Then a user could start up the service multiple times with different configurations, e.g.:

```
kdb set /sw/elektra/snippets/#0/current/cppcms/service/ip 127.0.0.1
kdb set /sw/elektra/snippets/#0/test/cppcms/service/ip 127.0.0.2
kdb rest-backend   # run default (current) with 127.0.0.1
kdb rest-backend --profile test  # run test config with 127.0.0.2
```

And there is even a profile plugin which allows you to do the same without command-line arguments, have a look at src/plugins/profile/README.md

markus2330 replied:
> About the "-"... I attempted to use the tool name, but in this case I need to use something static I guess?

You could just get rid of - in a dynamic way (restbackend).

qt-gui also uses qtgui in its config. But you should also consider renaming the whole thing (can be done after the PR), neither rest, backend, rest-backend nor restbackend is very appealing ;)

Namoshek replied:
Then I'll use the regex "[^a-zA-Z]" to sanitize the tool name. Renaming later should be possible, the cmake var `${tool}` should be used in every place where it matters.

For the profile stuff: I'll prepare it so that we could add it later and that "current" is the default for the moment.

Namoshek commented on src/tools/rest-backend/authentication_application.cpp:
```diff
@@ -316,4 +304,46 @@ model::User AuthenticationApp::getCurrentUser (cppcms::http::request & request)
+ * @param user a user model
+ * @return a jwt token as string
+ */
+std::string AuthenticationApp::buildJWT (cppcms::http::response & resp, const model::User & user) const
```
I actually had this improvement already in mind. Let's make a separate ticket for possible improvements for the future, if I get bored. ;)

markus2330 commented:
> I hope my Elektra integration makes sense!

Yes, nicely done. Only the regex was a bit surprising. The class "Config" needs some rework though, its not a good idea to have everything global and unprotected.

Namoshek commented on src/tools/rest-backend/README.md:
```diff
@@ -0,0 +1,191 @@
+The service itself offers quite some configuration options as well.
+In detail, the options (without the base key `/sw/elektra/@tool@/backend`) are:
+
+- `api_specification/raw` (string): A link to the blueprint describing the API.
```
Can I put this into a separate file that gets somehow mounted/imported during installation (+ maybe added to the readme through `configure_file`)? I'd like to provide a specification, but to be honest, I haven't found out yet how exactly it is done!

Namoshek commented on src/tools/rest-backend/root_application.cpp:
```diff
@@ -92,14 +100,75 @@ void RootApp::version ()
+
+		if (error)
+		{
+			// in case we could not retrieve the run-time version,
```
Where does the Elektra logger write to? And is there a particular reason why we do not have a convenience macro for ELEKTRA_LOG_ERROR?

markus2330 replied:
There are such macros, `ELEKTRA_LOG` is such a macro. See doc/CODING.md for explanations and src/libs/elektra/log.c for details.

Namoshek replied:
I think I wasn't clear enough. [kdblogger.h](../blob/master/src/include/kdblogger.h) enumerates ERROR as log level, but there is no macro for it directly available, only for the other four log levels.

markus2330 replied:
Yes, there is a reason: within the lib ERROR is reserved for `ELEKTRA_ASSERT`, otherwise there is no error (API misuse is the only error and it should be only triggered via assertions). I clarified it in 1cc8af24be035bafa0bf623bc942e3efff8b4f9d

You are right, however, for your scenario ERROR would be useful. But it is trivial to write, see bottom of src/include/kdblogger.h, just add the macro definitions in one of your headers.

Namoshek commented:
By the way, we need to install LibJWT on the build server, the backend currently does not get built as far as I see.

markus2330 commented:
> LibJWT

I think its best if we directly set up an agent and build job that fully deploys and runs the whole rest service. Please contact me by e-mail for login data and infos for which parts you can set up (jenkins, agent,..).

Namoshek commented:
Now everything should be ready to merge. :)

markus2330 commented:
Great job!

markus2330 commented:
With 287 comments, this is the most discussed PR.
